Automorphism groups for polytopes

lkastner opened the pull request.

lkastner committed:
PolyhedralGeometry: Use proper indexing shift

lkastner committed:
PolyhedralGeometry: First version of automorphism_group

lkastner committed:
PolyhedralGeometry: Streamline automorphism functions

lkastner committed:
PolyhedralGeometry: Refine docs of automorphism_group

lkastner committed:
Combinatorics: automorphisms of graphs should be actual groups

thofma commented on src/PolyhedralGeometry/Groups.jl:
```diff
@@ -164,7 +164,16 @@ end
 @doc Markdown.doc"""
     automorphism_group(P::Polyhedron; type = :combinatorial)
```
How about an `:action` keyword argument which could be `:vertex` or `:facet`? If this makes no sense or does not work out in practice, just discard my comment.

lkastner replied:
Let me know whether I understood you right.

thofma replied:
Yes, this is what I meant.

lkastner committed:
PolyhedralGeometry: Some functions should be hidden

lkastner committed:
PolyhedralGeometry: Test was wrong

lkastner committed:
PolyhedralGeometry: Further streamlining of automorphism

- introduce `action` parameter, defining what we want to act on,
  e.g. `:on_vertices` or `:on_facets`
- introduce `automorphism_group(IncidenceMatrix)`
- `linear_symmetries` now works properly, despite bugs in polymake

lkastner committed:
PolyhedralGeometry: Polish documentation of automorphism_group

lkastner commented:
FYI: This PR is waiting for
https://github.com/oscar-system/Polymake.jl/pull/404
to get merged.

benlorenz commented on src/PolyhedralGeometry/Groups.jl:
````diff
@@ -112,9 +109,218 @@ julia> length(elements(G))
+```
+"""
+function automorphism_group_generators(P::Polyhedron; type = :combinatorial, action = :all)
+    pm_object(P).BOUNDED || throw(ArgumentError("Automorphism groups not supported for unbounded polyhedra."))
````
```suggestion
    is_bounded(P) || throw(ArgumentError("Automorphism groups not supported for unbounded polyhedra."))
```

benlorenz commented on src/PolyhedralGeometry/Groups.jl:
```diff
@@ -112,9 +109,218 @@ julia> length(elements(G))
+
+
+function _linear_symmetries_generators(P::Polyhedron; action = :all)
     if pm_object(P).BOUNDED
```
```suggestion
    if is_bounded(P)
```

benlorenz commented on src/PolyhedralGeometry/Groups.jl:
```diff
@@ -70,11 +70,8 @@ julia> length(elements(G))
-        throw(ArgumentError("Symmetry groups currently supported for bounded polyhedra only"))
-    end
+    result = automorphism_group(P; type=:combinatorial)
+    return result[:on_vertices]
```
```suggestion
    return automorphism_group(P; type=:combinatorial, action=:on_vertices)
```

benlorenz commented on src/Combinatorics/Graphs.jl:
```diff
@@ -542,16 +542,34 @@ Return generators of the automorphism group of the graph `g`.
     pmg = pm_object(g);
     result = Polymake.graph.automorphisms(pmg)
-    return [[x+1 for x in a] for a in result]
+    return pm_arr_arr_to_group_generators(result)
```
```suggestion
    return _pm_arr_arr_to_group_generators(result)
```

lkastner committed:
Apply suggestions from code review

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

benlorenz commented on src/Combinatorics/Graphs.jl:
````diff
@@ -542,16 +542,34 @@ Return generators of the automorphism group of the graph `g`.
+```
+"""
+function automorphism_group(g::Graph{T}) where {T <: Union{Directed, Undirected}}
+    return gens_to_group(automorphism_group_generators(g))
````
```suggestion
    return _gens_to_group(automorphism_group_generators(g))
```

lkastner committed:
Update src/Combinatorics/Graphs.jl

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

lkastner committed:
Bump Polymake.jl compat, repair example

lkastner committed:
PolyhedralGeometry: Replace deprecated vf_group in tests
